net/speedtest/speedtest_server.go

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) 2021 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package speedtest
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"fmt"
+	"net"
+	"time"
+)
+
+// Serve starts up the server on a given host and port pair. It starts to listen for
+// connections and handles each one in a goroutine. Because it runs in an infinite loop,
+// this function only returns if any of the speedtests return with errors, or if the
+// listener is closed.
+func Serve(l net.Listener) error {
+	for {
+		conn, err := l.Accept()
+		if opErr, ok := err.(*net.OpError); ok {
+			if opErr.Err.Error() == listenerClosedErr {
```

Comment:
I think you want to use errors.Is here:

if errors.Is(err, net.ErrClosed) { // assuming that this is net.ErrClosed?
  return nil
}
if err != nil {
  return err
}